packages/agent/src/canisterStatus/index.ts

```diff
@@ -0,0 +1,241 @@
+import { Principal } from '@dfinity/principal';
+import { HttpAgent, HttpAgentOptions, Cbor, Certificate, toHex } from '..';
+
+/**
+ * Types of an entry on the canisterStatus map.
+ * An entry of null indicates that the request failed, due to lack of permissions or the result being missing.
+ */
+export type Status = string | ArrayBuffer | Date | ArrayBuffer[] | Principal[] | bigint | null;
+
+/**
+ * Interface to define a custom path. Nested paths will be represented as individual buffers, and can be created from text using {@link TextEncoder}
+ */
+export interface CustomPath {
+  key: string;
+  path: ArrayBuffer[];
+  decodeStrategy: 'cbor' | 'hex' | 'leb128' | 'raw';
+}
+
+/**
+ * Interface to request metadata from the icp:public or icp:private sections.
+ * Similar to {@link CustomPath}, but accepts a simple string argument.
+ * Private metadata will require the ${@link Identity} used by the ${@link HttpAgent} will need to be requested using an identity that controlls the canister.
+ */
+export interface MetaData {
+  kind: 'medadata';
+  key: string;
+  path: string | ArrayBuffer;
+  decodeStrategy: 'cbor' | 'hex' | 'leb128' | 'raw';
+}
+
+/**
+ * Pre-configured fields for canister status paths
+ */
+export type Path =
+  | 'time'
+  | 'controllers'
+  | 'subnet'
+  | 'module_hash'
+  | 'candid'
+  | MetaData
+  | CustomPath;
+
+export type StatusMap = Map<Path | string, Status>;
+
+export type CanisterStatusOptions = {
+  canisterId: Principal;
+  paths?: Path[] | Set<Path>;
+  agentOptions?: HttpAgentOptions;
+  agent?: HttpAgent;
+};
+
+/**
+ *
+ * @param {CanisterStatusOptions} options {@link CanisterStatusOptions}
+ * @param {CanisterStatusOptions['canisterId']} options.canisterId {@link Principal}
+ * @param {CanisterStatusOptions['paths']} options.paths {@link Path[]}
+ * @param {CanisterStatusOptions['agentOptions']} options.agentOptions {@link HttpAgentOptions} optional configuration for the agent used to make the request
+ * @param {CanisterStatusOptions['agent']} options.agent {@link HttpAgent} optional authenticated agent to use to make the canister request. Useful for accessing private metadata under icp:private
+ * @returns {Status} object populated with data from the requested paths
+ * @example
+ * const status = await canisterStatus({
+ *   paths: ['controllers', 'candid'],
+ *   ...options
+ * });
+ *
+ * const controllers = status.get('controllers');
+ */
+export const request = async (options: {
+  canisterId: Principal;
+  paths?: Path[] | Set<Path>;
+  agentOptions?: HttpAgentOptions;
+  agent?: HttpAgent;
+}): Promise<StatusMap> => {
+  const { canisterId, agentOptions, agent, paths } = options;
+
+  const uniquePaths = [...new Set(paths)];
+
+  // Set up agent
+  const effectiveAgent = agent ?? new HttpAgent(agentOptions);
+  if (options.agentOptions?.host && !options.agentOptions.host.startsWith('https://ic0.app')) {
+    await effectiveAgent.fetchRootKey();
+  }
```

Comment:
1. If an agent is passed in, I don't think we should ever call fetchRootKey on it.  Instead require the caller to have already set up their agent.
3. Based on this adjustment in the HttpAgent constructor, this comparison doesn't look safe, for example if the host is `<canister-id>.ic0.app`

https://github.com/dfinity/agent-js/blob/main/packages/agent/src/agent/http/index.ts#L186

```
    // Rewrite to avoid redirects
    if (this._host.hostname.endsWith(IC0_SUB_DOMAIN)) {
      this._host.hostname = IC0_DOMAIN;
    }
```

Can we make the caller explicitly indicate that they want to fetch the root key?  If not, would you consider removing the convenience of automatically creating an HttpAgent (so: retain `agent` parameter, remove `agentOptions` parameter)

Reply:
agreed. It's a little more cumbersome, but passing an agent in does resolve the potential inconsistency. We also don't manage replica-based behavior anywhere else in the agent, so we shouldn't do it here. I'll remove the convenience constructor